lib/Dialect/TritonGPU/Transforms/Pipeliner/TC05MMAPipeline.cpp

```diff
@@ -593,6 +593,31 @@ void createBarrierAndWaitOps(IRRewriter &builder, scf::ForOp forOp,
   annotateWithPipelineStage(builder, info.phase.getDefiningOp(), 0);
 }
 
+bool isSafeToPipeline(ttng::TCGen5MMAScaledOp scaledDot) {
+  auto getNumUsers = [](Value value) {
+    return std::distance(value.user_begin(), value.user_end());
+  };
+
+  auto isCopiedByTMEMCopy = [=](Value scale) {
+    if (getNumUsers(scale) != 2) {
+      // MMA and TMEM copy must be the only users
+      return false;
+    }
+
+    for (auto user : scale.getUsers()) {
+      if (isa<ttng::TMEMCopyOp>(user)) {
+        // If the scale is used by TMEM copy and the only other user is the
+        // scaled dot op, MMA pipelining is safe to apply.
+        return true;
+      }
+    }
+    return false;
```

Comment:
nit: maybe slightly more conservative?
```suggestion
    for (auto user : scale.getUsers()) {
      if (!isa<ttng::TMEMCopyOp, ttng::TCGen5MMAScaledOp>(user)) {
        // If the scale is used by TMEM copy and the only other user is the
        // scaled dot op, MMA pipelining is safe to apply.
        return false;
      }
    }
    return true;
```